### PR TITLE
Dashboard: Update dashboard grid space according to wordpress menu width

### PR DESCRIPTION
**Pull request by BrittanyIRL**
# Summary
Make sure dashboard is considering the total space available to stories regardless of wordpress left nav being collapsed or expanded 

## Relevant Technical Choices (originally)
- Initially I was only checking for the collapsed wordpress left nav, but when the menu's expanded it shrinks the dashboard's total available space. Since this is outside of our jurisdiction with the dashboard app, need to control this by watching for clicks on the menu and then recalculate the space available for the pageSize getting used in grid 
- Setting wordpress nav element `#adminmenu` as a ref, setting .current immediately to the element because we don't need it to cause any extra renders. 
- New `useLayoutEffect` to have an eventListener on nav for clicks. Debounce clicks and set width off of that instance of current.offsetWidth for nav.
- New `useLayoutEffect` to update the available container space with new wpMenuWidth when it changes. 
- wpMenuWidth defaults to 0

## Updated PR 
- this escalated from a 2 file change pr to 22 file changes - here is why: 
- Pascal made the request that this util try not to rely on wordpress and made me aware of a custom hook in the editStory directory called `useResizeEffect` that he'd rather get used. Originally, this usePagePreviewSize util was just a small thing to get the story sizes so we could render them - it turned out we needed that size for the grid not just the stories and need that util to be watching for the size available to get the proper size in pixels of stories in a grid setting as well as isolated. Originally doing this with just the one watcher for the container size was enough, but that didn't take into consideration the dashboard available width when the wordpress menus were expanded. 
- Implementing the `useResizeEffect` hook drastically reduced the checks we need to handle to get the pageSize and simplified the code so I think this update was worthwhile 
- While refactoring this code, I figured it was a good time to also clean up how many times the util is getting used and passing the pageSize into the grid elements that need it (namely the preview since it needs pixel size to overlay properly and prevent pageSize getting called repeatedly - this is where all of the extra file diff comes from - I had to update storybook and tests tied to this structure. 
- It's also important to note that we still have to tie our resizeEffect to a ref based on `wpbody` because this id exists outside of just the dashboard it can tell when the wordpress part of the page as a whole is also resized (meaning viewport AS WELL as menus expanding or collapsing). I spent some time yesterday experimenting with IDs inside of the dashboard app to keep it contained and it just doesn't work. This is something that will have to change when the dashboard is pulled out of wordpress, but it's 1 line to change and there's a constant in `pageStructure` making it pretty clear. 

## User-facing changes
- Grid items should now fill the total grid space regardless of if wordpress menu is expanded or collapsed

wp menu toggle resize

## Testing Instructions
- Toggle the wordpress menu open and closed and see that the grid updates the item size to fill the available space. 
- See that [storybook/myStories/content](http://localhost:59746/?path=/story/dashboard-components-mystories-content--default) works still (it has no wordpress menu, so this is a good spot to see that 0 come into play)
- see that list and grid views are as expected 

---

Fixes #

### CODE REVIEW

**Review by carloskelly13 — CHANGES_REQUESTED** [review]
I am getting `evaluating 'wpLeftNav.current.offsetWidth'` when I go to `myStores > Content > Default` in storybook.

**Comment by BrittanyIRL** [comment]
> I am getting `evaluating 'wpLeftNav.current.offsetWidth'` when I go to `myStores > Content > Default` in storybook.

weird! i did not get that. let me go back and fix that.

File: assets/src/dashboard/utils/usePagePreviewSize.js [hunk 1]
```diff
@@ -110,6 +118,24 @@ export default function usePagePreviewSize(options = {}) {
...
+    if (thumbnailMode || !wpLeftNav.current) {
+      return () => {};
+    }
+
+    const handleWpMenuToggle = () => debounceWpMenuToggle();
+
+    wpLeftNav.current.addEventListener('click', handleWpMenuToggle);
+    return () => {
+      wpLeftNav.current.removeEventListener('click', handleWpMenuToggle);
+    };
```
Comment (maxyinger) [inline assets/src/dashboard/utils/usePagePreviewSize.js]: nit: I think it's fine to just pass in `debounceWpMenuToggle` instead of renaming it here. 

Also:
```
const handleWpMenuToggle = () => debounceWpMenuToggle();
```
is the same as
```
const handleWpMenuToggle = debounceWpMenuToggle;
```
if you did want it renamed before you passed it in to the add listener

File: assets/src/dashboard/utils/usePagePreviewSize.js [hunk 2]
```diff
@@ -70,25 +69,34 @@ const sizeFromWidth = (
...
+    return innerWidth - wpMenuWidth - DASHBOARD_LEFT_NAV_WIDTH;
   } else {
-    return innerWidth - WP_LEFT_NAV_WIDTH;
+    return innerWidth - wpMenuWidth;
   }
 };
 
 export default function usePagePreviewSize(options = {}) {
+  const wpLeftNav = useRef(document.getElementById('adminmenu'));
+  wpLeftNav.current = document.getElementById('adminmenu');
```
Comment (maxyinger) [inline assets/src/dashboard/utils/usePagePreviewSize.js]: just a question cus I don't know the answer here. were you getting cases where `#adminmenu` wasn't mounted before this hook was run?

like does this work without this line below?
```
wpLeftNav.current = document.getElementById('adminmenu');
```

File: assets/src/dashboard/utils/usePagePreviewSize.js [hunk 3]
```diff
@@ -70,25 +69,34 @@ const sizeFromWidth = (
...
   );
 
   const [debounceAvailableContainerSpace] = useDebouncedCallback(() => {
-    setAvailableContainerSpace(getTrueInnerWidth());
+    setAvailableContainerSpace(getTrueInnerWidth(wpMenuWidth));
+  }, 250);
+
+  const [debounceWpMenuToggle] = useDebouncedCallback(() => {
+    setWpMenuWidth(wpLeftNav.current.offsetWidth);
   }, 250);
```
Comment (maxyinger) [inline assets/src/dashboard/utils/usePagePreviewSize.js]: one thing to note, and it should be fine perf wise for now, but can always cycle back to it later if we ever start having issues, but this hook is adding a new `onClick` listener every time it's instantiated. So if we ever have it rendered multiple times, we're making the DOM take these measurements multiple times to get the same result.

I don't think this hook is used in a situation like each item of a grid, but it it was, it's something to note that it will measure `wpLeftNav` `n` times for each time the hook is rendered

File: assets/src/dashboard/utils/usePagePreviewSize.js [hunk 4]
```diff
@@ -110,6 +118,24 @@ export default function usePagePreviewSize(options = {}) {
     availableContainerSpace,
   ]);
 
+  useLayoutEffect(
```
Comment (maxyinger) [inline assets/src/dashboard/utils/usePagePreviewSize.js]: Not sure if I'm seeing the need for layout effects here, but guess it doesn't hurt.

I think the order of events will fire like this with this `useLayoutEffect`, but not 100% sure:
```
// -> button clicked 
// -> wait 250ms for button to not be clicked again
// -> fire debounced listener which fires `setWpMenuWidth`
// -> render caused by `wpMenuWidth` update
// -> useLayoutEffect fires before DOM paint calling `setAvailableContainerSpace`
// -> DOM paints here
// -> render caused by `availableContainerSpace` update
// -> final DOM paint
```
I really think `useLayoutEffect` is only useful if you're getting or setting an DOM node's size directly in the effect and need this to be coordinated before a specific paint, but I still don't think this is even needed a lot of time in a lot of animation stuffs

File: assets/src/dashboard/utils/usePagePreviewSize.js [hunk 5]
```diff
@@ -110,6 +118,24 @@ export default function usePagePreviewSize(options = {}) {
...
+    () => setAvailableContainerSpace(getTrueInnerWidth(wpMenuWidth)),
+    [wpMenuWidth]
+  );
+
+  useLayoutEffect(() => {
```
Comment (maxyinger) [inline assets/src/dashboard/utils/usePagePreviewSize.js]: This is another case where `useLayoutEffect` doesn't hurt anything, but also doesn't add anything.
Here's the difference in order of events if you were to use this as an effect vs a layout effect:

**Layout Effect**
```
// -> render caused by first run or deps updating
// -> layout effect adds event listener before DOM paints
// -> DOM paints
```

**Effect**
```
// -> render caused by first run or deps updating
// -> DOM paints
// -> Effect adds event listener
```

In either case, the user is never gonna be able to click between a render and a paint. You could think of this being like:
```
const syncAddListener = (listener) => {
  elem.addEventListener('click', listener);
};
syncAddListener(() => console.log('click from sync added Fn'));

// vs
const asyncAddListener = (listener) => {
  setTimeout(() => elem.addEventListener('click', listener), 0);
};
asyncAddListener(() => console.log('click from async added Fn'));
```
with either one, the user's never going to be able to click fast enough for `"click from sync added Fn"` to be logged and not see `"click from async added Fn"` to be logged.

**Review by maxyinger — APPROVED** [review]
Good work!

I think right now we're using a lot of metrics to infer the size of the content area and infer when it changes. I think this is fine for now, but if we notice this code balooning in complexity from edge cases we missed, might be worth just adding a resize observer to the content area to centralize all this logic. 

I think everything's working great now tho! just something to keep an eye on if we find ourselves continually tuning this.

Also ran through your cases where `useLayoutEffect` was used. Moreso going over it for myself cus I always get mixed up on what it actually does, but hopefully there's something helpful in there to reason about it and it's not just me rambling at you about stuff you already know lol.

Anyway, good work! code runs great.

**Review by maxyinger — APPROVED** [review]
^^^ nvmd review went through earlier

File: assets/src/dashboard/utils/usePagePreviewSize.js (hunk 1 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/utils/usePagePreviewSize.js]: ah yup, had just copied my earlier pattern from the other useEffect. I'll fix this!

**Comment by BrittanyIRL** [comment]
> Also ran through your cases where `useLayoutEffect` was used. Moreso going over it for myself cus I always get mixed up on what it actually does, but hopefully there's something helpful in there to reason about it and it's not just me rambling at you about stuff you already know lol.

no way this was great! I think writing out the flow and order of hooks is maybe a good thing to get in the habit of on PRs (for myself) to walk through this better. The actual use cases for hooks vs how the info about them can be retained from docs are different. I went ahead and swapped out for useEffect.

File: assets/src/dashboard/utils/usePagePreviewSize.js (hunk 2 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/utils/usePagePreviewSize.js]: Yes! it does work without because adminmenu is outside of our Dashboard app completely. Thank you for questioning this :D Updated

**Comment by BrittanyIRL** [comment]
> I am getting `evaluating 'wpLeftNav.current.offsetWidth'` when I go to `myStores > Content > Default` in storybook.

Boo! This is fixed now. It was because I had been experimenting with whether useRef was better for the menu. I went with yes but in my update back to ref.current i missed the initial check when setting the wpMenuWidth to check current exists which was causing the issue. Sorry about that little delay!

File: assets/src/dashboard/utils/usePagePreviewSize.js [hunk 6]
```diff
@@ -98,18 +104,32 @@ export default function usePagePreviewSize(options = {}) {
...
+    wpMenuWidth,
+  ]);
+
+  useEffect(() => {
+    if (thumbnailMode || !wpLeftNav.current) {
+      return () => {};
+    }
+    const wpLeftNavCurrent = wpLeftNav.current;
+
+    wpLeftNavCurrent.addEventListener('click', debounceWpMenuToggle);
```
Comment (swissspidy) [inline assets/src/dashboard/utils/usePagePreviewSize.js]: I don't like how tightly coupled this code is to WordPress. Especially as it relies on detecting clicks on this toggle button. It'd be more reliable to check for the `folded` class on the `<body>`. However, it is still coupled to WordPress.

We went down that road for the editor already and ended up implementing a [`useResizeEffect`](https://github.com/google/web-stories-wp/blob/200897d5d911bf6aa20faf777b293a8f6d28f78f/assets/src/edit-story/utils/useResizeEffect.js) hook to detect how much space is available. Here's how we currently use it in the editor's carousel:

https://github.com/google/web-stories-wp/blob/master/assets/src/edit-story/components/canvas/carousel/index.js#L221-L231

Bonus: if you adopt this hook in the dashboard, you could also use it in `ToggleButtonGroup`.

**tl;dr**: let's avoid relying on WP environment specifics.

File: assets/src/dashboard/app/views/myStories/content/index.js [hunk 7]
```diff
@@ -80,7 +80,7 @@ function Content({
                   stories={stories}
                   tags={tags}
                   users={users}
-                  viewStyle={view.style}
+                  view={view}
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/views/myStories/content/index.js:83]: passing whole view object in now instead of just the style so we can easily reference pageSize

File: assets/src/dashboard/components/cardGrid/index.js [hunk 8]
```diff
@@ -23,10 +23,10 @@ import styled from 'styled-components';
 /**
  * Internal dependencies
  */
-import usePagePreviewSize from '../../utils/usePagePreviewSize';
+import { PageSizePropType } from '../../types';
 
-const DashboardGrid = styled.div`
-  ${({ columnHeight, columnWidth, theme }) => `
+const DashboardGrid = styled.div(
+  ({ columnHeight, columnWidth, theme }) => `
```
Comment (carloskelly13) [inline assets/src/dashboard/components/cardGrid/index.js:29]: Since we are passing ?. optionals here we should provide some kind of default value so `undefined` or `null` `px` isn't the grid gaps.

File: assets/src/dashboard/utils/index.js [hunk 9]
```diff
@@ -24,3 +24,4 @@ export { default as useDashboardResultsLabel } from './useDashboardResultsLabel'
 export { default as usePagePreviewSize } from './usePagePreviewSize';
 export { default as useStoryView } from './useStoryView';
 export { default as throttleToAnimationFrame } from './throttleToAnimationFrame';
+export { default as useResizeEffect } from '../../edit-story/utils/useResizeEffect';
```
Comment (BrittanyIRL) [inline assets/src/dashboard/utils/index.js:27]: Thoughts about this structure for when we reference edit-story utils instead of copy and paste? I think this makes it clear where the codeshare is.

File: assets/src/dashboard/utils/usePagePreviewSize.js [hunk 10]
```diff
@@ -54,6 +50,11 @@ const sizeFromWidth = (
   if (respectSetWidth) {
     return { width, height: width / PAGE_RATIO };
   }
+
+  if (bp === 'desktop') {
+    availableContainerSpace -= DASHBOARD_LEFT_NAV_WIDTH;
```
Comment (BrittanyIRL) [inline assets/src/dashboard/utils/usePagePreviewSize.js:55]: turns out it's better if we base this math on the straight breakpoint not the inner width because the inner width changes based on if the wordpress menu is collapsed wheras the dashboard nav menu is based on the full viewport width. Doing this instead of `getTrueInnerWidth` helps _alot_

File: assets/src/dashboard/utils/usePagePreviewSize.js (hunk 6 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/utils/usePagePreviewSize.js]: thanks for this note, @swissspidy! I refactored to use the approach you suggested - it was something we were hesitant to spend the time on but i appreciate the desire for continuity. I think it turned out nice if you wanna take another peek. I did update the useResizeEffect slightly to check that the ref exists. before grabbing .current and in cases where that node is false returning undefined to remove that console warning about not returning null.

File: assets/src/dashboard/utils/usePagePreviewSize.js [hunk 11]
```diff
@@ -68,47 +69,38 @@ const sizeFromWidth = (
...
-  } else {
-    return innerWidth - WP_LEFT_NAV_WIDTH;
-  }
-};
-
 export default function usePagePreviewSize(options = {}) {
   const { thumbnailMode = false, isGrid } = options;
+  // When the dashboard is pulled out of wordpress this id will need to be updated.
+  // For now, we need to grab wordpress instead because of how the app's rendered
+  const dashboardContainerRef = useRef(document.getElementById(WPBODY_ID));
```
Comment (carloskelly13) [inline assets/src/dashboard/utils/usePagePreviewSize.js:76]: This will be a nice way to quickly adjust that.

Comment (BrittanyIRL) [reply assets/src/dashboard/utils/usePagePreviewSize.js:76]: that's what i was thinking too, i'm glad this makes sense

File: assets/src/dashboard/components/cardGrid/index.js (hunk 8 above)
Comment (BrittanyIRL) [reply assets/src/dashboard/components/cardGrid/index.js:29]: I opted to take out `?.` optionals here, the pageSize is _always_ going to be passed through. And if it's not, this value needs to originate higher up so that the other places pageSize is getting used from the same context consumer are in sync. I had added the check here earlier when testing out storybook to get to something else and missed this. Sorry about that.

**Review by carloskelly13 — APPROVED** [review]
